dataset/src/main/scala/frameless/TypedColumn.scala

```diff
@@ -471,6 +471,31 @@ abstract class AbstractTypedColumn[T, U]
   def cast[A: TypedEncoder](implicit c: CatalystCast[U, A]): ThisType[T, A] =
     typed(self.untyped.cast(TypedEncoder[A].catalystRepr))
 
+  /**
+    * An expression that returns a substring
+    * {{{
+    *   df.select(df('a).substr(0, 5))
+    * }}}
+    *
+    * @param startPos starting position
+    * @param len length of the substring
+    */
+  def substr(startPos: Int, len: Int)(implicit ev: U =:= String): ThisType[T, String] =
+    typed(self.untyped.substr(startPos, len))
+
+  /**
+    * An expression that returns a substring
+    * {{{
+    *   df.select(df('a).substr(df('b), df('c)))
+    * }}}
+    *
+    * @param startPos expression for the starting position
+    * @param len expression for the length of the substring
+    */
+  def substr[TT, W](startPos: ThisType[TT, Int], len: ThisType[TT, Int])
```

Comment:
I think this is the first columns method we have that involves 3 columns. The way you wrote it here, with `TT` used in both `startPos` and `len`, you are forcing these two columns to come from the same dataset. Something like the following wouldn't typecheck:

```scala
ds1.joins(ds2)(ds1('a) === ds1('a).sustr(ds1('b), ds2('c))
```

I know it's a contrive example, but to make the above working you could need something like the following:

```scala
def substr[TT1, TT2, W1, W2](startPos: ThisType[TT1, Int], len: ThisType[TT2, Int])
  (implicit
    i0: U =:= String,
    i1: With.Aux[T, TT1, W1],
    i2: With.Aux[W1, TT2, W2]
  ) = ...
```

Reply:
make sense, I will fix it. thanks